loki/transformations/temporaries/stack_allocator.py

```diff
@@ -0,0 +1,1372 @@
+        stack_dict : dict
+            dict that maps dtype and kind to an expression for the required stack size
+        successors : list of :any:`Item`
+            The items corresponding to successor routines called from :data:`routine`
+        """
+
+        stack_vars = []
+        stack_args = []
+        stack_arg_dict = {}
+        pragma_string = ''
+        assignments = []
+        for dtype in stack_dict:
+            for kind in stack_dict[dtype]:
+
+                # start arguments integer names in kernels with 'K'
+                stack_size_name = self._get_stack_int_name('K', dtype, kind, 'STACK_SIZE')
+                stack_size_var = self._get_int_var(name=stack_size_name, scope=routine,
+                                                   type=self._get_int_type(intent='IN'))
+
+                # local variables start with 'J'
+                stack_used_arg_name = self._get_stack_int_name('JD', dtype, kind, 'STACK_USED')
+                stack_used_arg = self._get_int_var(name=stack_used_arg_name, scope=routine,
+                                                   type=self._get_int_type(intent='INOUT'))
+                stack_used_name = self._get_stack_int_name('J', dtype, kind, 'STACK_USED')
+                stack_used_var = self._get_int_var(name=stack_used_name, scope=routine)
+                assignments += [Assignment(lhs=stack_used_var, rhs=stack_used_arg)]
+
+                # create the stack variable and its type with the correct shape
+                shape = (stack_size_var,)
+                stack_var = self._get_stack_var(routine, dtype, kind)
+                stack_type = stack_var.type.clone(shape=as_tuple(shape), target=True)
+                stack_var = stack_var.clone(type=stack_type)
+
+                # pass on the stack variable from stack_used + 1 to stack_size
+                # pass stack_size - stack_used to stack size in called kernel
+                arg_dims = (RangeIndex((None, None)),)
+                stack_arg_dict.setdefault(dtype, {})
+                stack_arg_dict[dtype][kind] = (stack_size_var, stack_var.clone(dimensions=arg_dims), stack_used_var)
+
+                # create stack_vars pair
+                stack_args += [stack_size_var,
+                               stack_var.clone(dimensions=stack_type.shape, type=stack_var.type.clone(contiguous=True)),
+                               stack_used_arg]
+                stack_vars += [stack_used_var]
+                pragma_string += f'{stack_var.name}, '
+
+        if pragma_string:
+            # remove ', '
+            pragma_string = pragma_string[:-2].lower()
+
+            present_pragma = None
+            acc_pragmas = [p for p in FindNodes(Pragma).visit(routine.body) if p.keyword.lower() == 'loki'] # acc
+            for pragma in acc_pragmas:
+                if pragma.content.lower().startswith('device-present'):
+                    present_pragma = pragma
+                    break
```

Comment:
Seems like this search could actually be combined:

```suggestion
            present_pragmas = [
                p for p in FindNodes(Pragma).visit(routine.body)
                if p.keyword.lower() == 'loki' and p.content.lower().startswith('device-present')
            ]
```